### PR TITLE
feat: add corner radius and window wrapping to rect_nodes

### PR DESCRIPTION
**Pull request by WillPower3309**

### CODE REVIEW

**Comment by WillPower3309** [comment]
For some reason the example isn't rounding its outer corners, will take a look tomorrow

File: include/scenefx/types/wlr_scene.h [hunk 1]
```diff
@@ -386,10 +387,16 @@ struct wlr_scene_rect *wlr_scene_rect_create(struct wlr_scene_tree *parent,
 		int width, int height, const float color[static 4]);
 
 /**
- * Change the width and height of an existing rectangle node.
+ * change the width and height of an existing rectangle node.
```
Comment (ErikReider) [inline include/scenefx/types/wlr_scene.h]: Intended change?

File: render/fx_renderer/fx_pass.c [hunk 2]
```diff
@@ -398,6 +398,21 @@ void fx_render_pass_add_rounded_rect(struct fx_gles_render_pass *pass,
...
+		pixman_region32_init(&clip_region);
+		pixman_region32_copy(&clip_region, options->clip);
+	} else {
+		pixman_region32_init_rect(&clip_region, box.x, box.y, box.width, box.height);
+	}
+	const struct wlr_box window_box = fx_options->window_box;
+	int window_corner_radius = fx_options->window_corner_radius;
+	pixman_region32_t window_region;
+	pixman_region32_init_rect(&window_region, window_box.x + window_corner_radius * 0.3, window_box.y + window_corner_radius * 0.3,
+			window_box.width - window_corner_radius * 0.6, window_box.height - window_corner_radius * 0.6);
```
Comment (ErikReider) [inline render/fx_renderer/fx_pass.c]: Make sure that the width and height aren't 0

File: render/fx_renderer/fx_pass.c [hunk 3]
```diff
@@ -456,18 +476,18 @@ void fx_render_pass_add_box_shadow(struct fx_gles_render_pass *pass,
...
 	} else {
 		pixman_region32_init_rect(&clip_region, box.x, box.y, box.width, box.height);
 	}
+	const struct wlr_box window_box = options->window_box;
+	int window_corner_radius = options->window_corner_radius;
 	pixman_region32_t window_region;
-	pixman_region32_init_rect(&window_region, window_box.x + options->window_corner_radius * 0.3, window_box.y + options->window_corner_radius * 0.3,
-			window_box.width - options->window_corner_radius * 0.6, window_box.height - options->window_corner_radius * 0.6);
+	pixman_region32_init_rect(&window_region, window_box.x + window_corner_radius * 0.3, window_box.y + window_corner_radius * 0.3,
+			window_box.width - window_corner_radius * 0.6, window_box.height - window_corner_radius * 0.6);
```
Comment (ErikReider) [inline render/fx_renderer/fx_pass.c]: Might as well change this here instead of my PR

File: tinywl/tinywl.c [hunk 4]
```diff
@@ -762,6 +777,8 @@ static void xdg_toplevel_destroy(struct wl_listener *listener, void *data) {
 	wl_list_remove(&toplevel->request_maximize.link);
 	wl_list_remove(&toplevel->request_fullscreen.link);
 
+	wlr_scene_node_destroy(&toplevel->border->node);
+	wlr_scene_node_destroy(&toplevel->shadow->node);
```
Comment (ErikReider) [inline tinywl/tinywl.c]: Isn't needed. Automatically called when its parent is destroyed, the line below :)

File: types/scene/wlr_scene.c [hunk 5]
```diff
@@ -1340,6 +1349,23 @@ static void scene_entry_render(struct render_list_entry *entry, const struct ren
...
+
+			transform_output_box(&window_box, data);
+
+			struct fx_render_rounded_rect_options rounded_rect_options = {
+				.base = rect_options.base,
+				.corner_radius = scene_rect->corner_radius,
+				.window_box = window_box,
+				.window_corner_radius = window_corner_radius,
+			};
+			fx_render_pass_add_rounded_rect(data->render_pass, &rounded_rect_options);
```
Comment (ErikReider) [inline types/scene/wlr_scene.c:1373]: Don't forget to break here

**Review by ErikReider — CHANGES_REQUESTED** [review]

**Comment by WillPower3309** [comment]
Need to get window resizing working in tinywl again

**Comment by WillPower3309** [comment]
Will add window wrapping to other rect nodes as a follow-up

**Pull request merged**